Sources/OktaAnalytics/OktaAnalytics.swift

```diff
@@ -102,12 +109,199 @@ public final class OktaAnalytics: NSObject {
+        */
+    public static func updateScenario(_ scenarioID: ScenarioID, _ propertySubject: (PassthroughSubject<Property, Never>?) -> Void) {
+        lock.readLock()
+        defer { lock.unlock() }
+        guard let scenario = Self.scenarios?[scenarioID] else {
+            assert(false, "startScenario should be called before updateScenario")
+            propertySubject(nil)
+            return
+        }
+        Self.providers.forEach {
+            $1.logger?.log(level: .debug, eventName: scenario.name, message: "\($0) Scenario \(scenario.name) Updated", properties: nil, file: #file, line: #line, funcName: #function)
+        }
+        propertySubject(scenario.eventStream)
+    }
+
+    /**
+        end sceanrio with ID. send properties to provider and clears local storage.
+
+        - Parameters:
+           - scenarioID: unique sceario ID returned from `startScenario(_ , _)` .
+           - eventDisplayName: event name to display on provider dashboard.
+        */
+    public static func endScenario(_ scenarioID: ScenarioID, eventDisplayName: Name) {
+        lock.readLock()
+        defer { lock.unlock() }
+
+        guard let scenario = Self.scenarios?[scenarioID] else {
+            assert(false, "No Scenario to end")
+            return
+        }
+
+        scenario.eventStream?.send(completion: .finished)
+
+        Self.trackEvent(eventDisplayName, withProperties: scenario.properties)
+        Self.scenarios?.removeValue(forKey: scenarioID)
+
+        let scenarioFetchRequest = NSFetchRequest<Scenario>(entityName: "Scenario")
+        scenarioFetchRequest.predicate = NSPredicate(format: "scenarioID CONTAINS %@", scenarioID)
+        do {
+            let scenarios = try Self.coreDataStack.managedContext.fetch(scenarioFetchRequest)
+            for scenario in scenarios {
+                Self.coreDataStack.managedContext.delete(scenario)
+            }
+        } catch {
+            assert(false, "Failed to fetch scenarios")
+        }
+        Self.coreDataStack.managedContext.saveContext()
+
+        let scenarioPropertiesFetchRequest = NSFetchRequest<ScenarioProperty>(entityName: "ScenarioProperty")
+        scenarioPropertiesFetchRequest.predicate = NSPredicate(format: "scenarioID CONTAINS %@", scenarioID)
+        do {
+            let scenarioProperties = try Self.coreDataStack.managedContext.fetch(scenarioPropertiesFetchRequest)
+            for scenarioProperty in scenarioProperties {
+                Self.coreDataStack.managedContext.delete(scenarioProperty)
+            }
+        } catch {
+            assert(false, "Failed to fetch scenario properties")
+        }
+        Self.coreDataStack.managedContext.saveContext()
+    }
+
+    /**
+        end sceanrio with scenario name. send all properties to provider and clears local storage.
+
+        - Parameters:
+           - name: scenario name
+        */
+    public static func endScenario(_ name: Name) {
```

Comment:
Why are we using name here, if we give a handle with id?

Reply:
This is to end scenarios with name, there might be multiple scenarios with same name. if interrupted before